lucene/core/src/java/org/apache/lucene/document/ShapeQuery.java

```diff
@@ -265,10 +265,20 @@ private Scorer getSparseScorer(final LeafReader reader, final Weight weight, fin
         final DocIdSetIterator iterator = new BitSetIterator(result, cost[0]);
         return new ConstantScoreScorer(weight, boost, scoreMode, iterator);
       }
-      final DocIdSetBuilder docIdSetBuilder = new DocIdSetBuilder(reader.maxDoc(), values, query.getField());
-      values.intersect(getSparseVisitor(query, docIdSetBuilder));
-      final DocIdSetIterator iterator = docIdSetBuilder.build().iterator();
-      return new ConstantScoreScorer(weight, boost, scoreMode, iterator);
+      if (values.getDocCount() << 2 < values.size()) {
```

Comment:
I think we should be careful with overflows, maybe divide instead of multiplying, ie.
```suggestion
      if (values.getDocCount() < (values.size() >>> 2)) {
```

Reply:
indeed that is safer